SCAutolib/controller.py

```diff
@@ -0,0 +1,375 @@
+                         f"to /etc/sssd/sssd.conf")
+        else:
+            if self.ipa_ca is None:
+                msg = "Can't proceed in configuration of IPA user because no " \
+                      "IPA Client is configured"
+                raise SCAutolibException(msg)
+            new_user = user.IPAUser(ipa_server=self.ipa_ca,
+                                    username=user_dict["name"],
+                                    pin=user_dict["pin"],
+                                    password=user_dict["passwd"],
+                                    card_dir=card_dir,
+                                    cert=user_dict["cert"],
+                                    key=user_dict["key"])
+
+            new_user.add_user()
+
+        new_card = None
+
+        if user_dict["card_type"] == "virtual":
+            hsm_conf = file.SoftHSM2Conf(
+                new_user.card_dir.joinpath("sofhtsm2.conf"),
+                new_user.card_dir)
+            hsm_conf.create()
+            hsm_conf.save()
+
+            new_card = card.VirtualCard(new_user)
+            new_card.softhsm2_conf = hsm_conf
+        else:
+            raise NotImplementedError("Other card type than 'virtual' does not "
+                                      "supported yet")
+        new_card.create()
+        new_user.card = new_card
+        return new_user
+
+    def enroll_card(self, user_: user.User):
+        """
+        Enroll the card of a given user with configured CA. If private key
+        and/or the certificate are not exists, new one's would be requested
+        from corresponding CA.
+
+        :param user_: User with a card to be enrolled.
+        """
+        logger.debug(f"Starting enrollment of the card for user {user_.username}")
+        if not user_.card:
+            raise SCAutolibException(f"Card for the user {user_.username} does "
+                                     f"not initialized")
+        if user_.cert is None:
+            # Creating a new private key makes sense only if the certificate
+            # doesn't exist yet
+            if user_.key is None:
+                user_.key = user_.card_dir.joinpath("private.key")
+                _gen_private_key(user_.key)
+
+            csr = user_.gen_csr()
+            if user_.cert is None:
+                user_.cert = user_.card_dir.joinpath(
+                    f"cert-{user_.username}.pem")
+
+            if isinstance(user_, user.IPAUser):
+                self.ipa_ca.request_cert(csr, user_.username, user_.cert)
+            else:
+                self.local_ca.request_cert(csr, user_.username, user_.cert)
+
+        user_.card.enroll()
+
+    def cleanup(self):
+        ...
+
+    def _validate_configuration(self, params: {} = None):
+        """
+        Validate schema of the configuration file. If some value doesn't present
+        in the config file, this value would be looked in the CLI parameters
```

Comment:
I'm wondering if we should prioritize CLI parameters over config file, I think it might make more sense

Reply:
Created an Issue #84